File: app/recursive_loop/rect_cover.py
# coding=utf-8
# author: Xiguang Liu<g10guang@gmail.com>
# 2018-04-27 11:26
# 题目描述：https://www.nowcoder.com/practice/72a5a919508a4251859fb2cfb987a0e6?tpId=13&tqId=11163&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking

import unittest


class Solution:
    def __init__(self):
        self.dp = {0: 0, 1: 1, 2: 2}

    def rectCover(self, number):
        """
        动态规划
        :param number:
        :return:
        """
        if number in self.dp:
            return self.dp[number]
        # 错误：f = self.rectCover(number-1) + self.rectCover(number-2)*2
        # 因为 self.rectCover(number-1) 已经包含了 self.rectCover(number-2) 的一种情况，这种情况值得琢磨
        f = self.rectCover(number-1) + self.rectCover(number-2)
        self.dp[number] = f
        return f


class TestCase(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_1(self):
        n = 3
        r = self.s.rectCover(n)
        self.assertEqual(r, 3)

File: app/recursive_loop/rect_cover.go
package main

/**
 * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
 *
 * @param number int整型
 * @return int整型
 */
// https://www.nowcoder.com/practice/72a5a919508a4251859fb2cfb987a0e6?tpId=13&tqId=11163&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking
func rectCover(number int) int {
	n_1 := 2
	n_2 := 1
	result := 0
	for i := 3; i <= number; i++ {
		result = n_1 + n_2
		n_2 = n_1
		n_1 = result
	}
	if number == 1 {
		return 1
	}
	if number == 2 {
		return 2
	}
	return result
}
